File: packages/earthsci-toolkit/src/staggering-rules-roundtrip.test.ts
/**
 * Round-trip tests for the §7.4 staggering_rules top-level schema (esm-15f).
 *
 * Loads the MPAS C-grid staggering fixture, serializes via save(), reloads,
 * and asserts the `staggering_rules` subtree survives the round-trip byte-
 * equivalently at the JSON level.
 */
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'fs'
import { join } from 'path'
import { load, save } from './index.js'

const fixturePath = join(__dirname, '../../../tests/grids/mpas_c_grid_staggering.esm')

describe('§7.4 staggering_rules top-level schema — round-trip', () => {
  it('preserves the mpas_c_grid_staggering rule across load/save/load', () => {
    const raw = readFileSync(fixturePath, 'utf-8')
    const loaded = load(raw)
    const serialized = save(loaded)
    const reloaded = load(serialized)
    const original = JSON.parse(raw) as { staggering_rules?: unknown }
    const after = (reloaded as unknown as { staggering_rules?: unknown }).staggering_rules
    expect(after).toBeDefined()
    expect(after).toEqual(original.staggering_rules)
  })

  it('exposes the rule fields on the typed model', () => {
    const raw = readFileSync(fixturePath, 'utf-8')
    const loaded = load(raw) as unknown as {
      staggering_rules?: Record<string, {
        kind: string
        grid: string
        edge_normal_convention?: string
        cell_quantity_locations?: Record<string, string>
      }>
    }
    expect(loaded.staggering_rules).toBeDefined()
    const rule = loaded.staggering_rules!['mpas_c_grid_staggering']
    expect(rule.kind).toBe('unstructured_c_grid')
    expect(rule.grid).toBe('mpas_cvmesh')
    expect(rule.edge_normal_convention).toBe('outward_from_first_cell')
    expect(rule.cell_quantity_locations?.u).toBe('edge_midpoint')
    expect(rule.cell_quantity_locations?.zeta).toBe('vertex')
  })
})

File: packages/esm-format-go/pkg/esm/staggering_rules_roundtrip_test.go
package esm

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// TestStaggeringRulesRoundtrip verifies the Go binding parses the §7.4
// top-level `staggering_rules` map (esm-15f) on the mpas_c_grid_staggering
// fixture and preserves it across a marshal / unmarshal cycle.
func TestStaggeringRulesRoundtrip(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}

	path := filepath.Join(repoRoot, "tests", "grids", "mpas_c_grid_staggering.esm")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}

	esmFile, err := LoadString(string(data))
	if err != nil {
		t.Fatalf("initial load: %v", err)
	}
	if len(esmFile.StaggeringRules) != 1 {
		t.Fatalf("expected 1 staggering rule, got %d", len(esmFile.StaggeringRules))
	}
	rule, ok := esmFile.StaggeringRules["mpas_c_grid_staggering"]
	if !ok {
		t.Fatalf("missing mpas_c_grid_staggering rule")
	}
	if rule.Kind != "unstructured_c_grid" {
		t.Errorf("kind: want unstructured_c_grid, got %q", rule.Kind)
	}
	if rule.Grid != "mpas_cvmesh" {
		t.Errorf("grid: want mpas_cvmesh, got %q", rule.Grid)
	}
	if rule.EdgeNormalConvention != "outward_from_first_cell" {
		t.Errorf("edge_normal_convention: want outward_from_first_cell, got %q",
			rule.EdgeNormalConvention)
	}
	if loc := rule.CellQuantityLocations["u"]; loc != "edge_midpoint" {
		t.Errorf("u should live at edge_midpoint, got %q", loc)
	}

	out, err := esmFile.ToJSON()
	if err != nil {
		t.Fatalf("marshal back: %v", err)
	}

	var orig, round map[string]interface{}
	if err := json.Unmarshal(data, &orig); err != nil {
		t.Fatalf("unmarshal original: %v", err)
	}
	if err := json.Unmarshal(out, &round); err != nil {
		t.Fatalf("unmarshal round-tripped: %v", err)
	}

	origRules := orig["staggering_rules"]
	roundRules := round["staggering_rules"]
	if !reflect.DeepEqual(origRules, roundRules) {
		t.Errorf("staggering_rules block did not round-trip.\n orig:  %#v\n round: %#v",
			origRules, roundRules)
	}
}
